README.md
<a href="https://www.nies.futbol/"><img
src="https://raw.githubusercontent.com/niesfutbol/hierarchical_review_mx/develop/static/logo_nies.png" align="right" width="256"
/></a>

# Template for Streamlit projects

## Step to use
### Local app
``` sh
docker-compose up -d app
```
Then http://localhost:8502/
### Develop Enviroment
``` sh
docker-compose run nies
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2023-08-20

### Added

### Fixed

### Changed

### Removed

[0.1.0]: https://github.com/niesfutbol/hierarchical_review_mx/compare/init...v0.1.0

tests/test_transformations.py
import hierarchical_review_mx as dt


def test_add_offset():
    augend = 1
    addend = 2
    expected = augend + addend
    obtained = dt.add_offset(augend, addend)
    assert expected == obtained


hr_MX_app.py
import pandas as pd
import streamlit as st
import requests
import hierarchical_review_plots as hrp


conn = requests.get("http://104.248.109.197:6969/v1/percentile_with_league_id")
larga = pd.DataFrame.from_dict(conn.json())
mp = pd.read_csv("static/minutes_played_23.csv")
names_teams = pd.read_csv("static/names_262_2022.csv")


def list_of_players_in_ws_and_as(longer, played_minutes):
    return [
        jugador
        for jugador in longer.Player.unique().tolist()
        if jugador in played_minutes.player.to_list()
    ]


"""
El objetivo de este análisis es caracterizar el estilo de los jugadores a partir de sus acciones
en el campo de juego.

En la primer pestaña encontrarás un resumen visual del desempeño de las jugadoras y jugadores de las
ligas mexicanas. Estas gráficas mostrarán distintas métricas dependiendo de la categoría a la que
pertenezca cada jugador.

En la segunda pestaña presentamos un resumen de los detalles técnicos del análisis. En NIES
esperamos que esta información sea de su interés.
"""

league_id_from_name = {"Liga MX": 262, "Liga Expansión": 263, "Liga MX Femenil": 673}
list_of_league = list(league_id_from_name.keys())
player, otro = st.tabs(["Jugador", "Explicación"])

with player:
    st.subheader("Gráficas de desempeño")
    league_name = st.selectbox("Selecciona una liga:", list_of_league)
    league_id = league_id_from_name[league_name]
    names_teams = pd.read_csv(f"static/names_{league_id}_2022.csv")
    list_of_teams = names_teams.names.to_list()
    team_name = st.selectbox("Selecciona un equipo:", list_of_teams)
    team_id = names_teams[names_teams.names == team_name]["ids"].to_list()[0]
    league_players = mp[(mp.league_id == league_id) & (mp.Team == team_name)]
    wy_players = league_players["Player"].unique().tolist()
    """
    Estas gráficas tienen un conjunto de métricas seleccionadas a partir de técnicas de inteligencia
    artificial.
    Cada barra representa la fuerza relativa del jugador en cada una de las métricas.
    La distancia que existe de la barra al centro indica el percentil comparado con la base de datos completa.

    Encontrarás la descripción completa en la nota: [Gráfica de desempeño de jugadores](https://www.nies.futbol/2023/07/grafica-de-desempeno-de-jugadores.html).
    """
    # ------------- game start ------------
    logo = {262: "logo_liga_mx", 263: "logo_expansion", 673: "logo_liga_mx_femenil"}
    radar_player = st.selectbox(f"Seleccciona un jugador:", wy_players)
    player_t = league_players[league_players.Player == radar_player]
    minutes_played = mp[mp.Player == radar_player]["Minutes played"].values[0]
    team_name = mp[(mp.Player == radar_player) & (mp.league_id == league_id)]["Team"].values[0]
    scotland_logo = logo[league_id]
    ac_milan_logo = f"logo_{team_id}"
    pizza_plot = hrp.make_bar_plot_player(
        larga[(larga.Player == radar_player) & (larga.league_id == league_id)],
        radar_player,
        minutes_played,
        team_name,
        league_logo=scotland_logo,
        team_logo=ac_milan_logo,
    )
    st.plotly_chart(pizza_plot, use_container_width=True)

with otro:
    st.subheader("Macro grupo")
    """
    Dividimos al total de jugadores de la base de datos en cinco grupos. Esta clasificación la
    hicimos a partir de técnicas de Inteligencia Artificial. Utilizamos más de 100 mediciones
    que tenemos de cada jugador. Nuestro algoritmo entrenado decidió cuáles eran las variables
    relevantes para agrupar jugadores dependiendo de sus números.

    Utilizamos la base completa que incluye los datos de cerca de 8,000 desempeños anuales, de
    difernetes ligas. Tenemos 3 temporadas de las principales ligas europeas y ligas mexicanas.
    Aquí tenemos dos restricciones, solo aparecen jugadores con más de 900 minutos disputados en una
    temporada y quitamos a los porteros.
    
    En la figura de abajo podemos ver los cinco macrogrupos en los que clasificamos a los jugadores.
    Cada punto representa el desempeño anual de algún jugador. Así, podríamos tener varios puntos
    para un solo jugador, un punto por cada temporada de la que tenemos registro.
    """
    st.image("static/figurita_soccerment.png")


st.markdown("Hecho con 💖 por [nies.futbol](https://nies.futbol)")


hierarchical_review_mx/polar_bar_plotly.py
import plotly.express as px


def make_bar_plot_player(larga, radar_player):
    fig = _set_up_bar_polar(larga, radar_player)
    return _update_bar_polar(fig)


def add_nies_logo(fig):
    fig.add_layout_image(
        dict(
            source="https://raw.githubusercontent.com/niesfutbol/hierarchical_review_mx/develop/static/logo_serie_a.png",
            xref="paper",
            yref="paper",
            x=0.9,
            y=1.05,
            sizex=0.2,
            sizey=0.2,
            xanchor="right",
            yanchor="bottom",
        )
    ).add_layout_image(
        dict(
            source="https://raw.githubusercontent.com/niesfutbol/hierarchical_review_mx/develop/static/logo_nies.png",
            xref="paper",
            yref="paper",
            x=0.05,
            y=0.05,
            sizex=0.2,
            sizey=0.2,
        )
    )
    return fig


def _set_up_bar_polar(larga, radar_player):
    player_t = larga[larga.Player == radar_player]
    fig = px.bar_polar(
        player_t,
        r="deciles",
        theta="variable",
        color="type_variable",
        title=f"Gráfica Radial de Barras Interactiva de {radar_player}",
    )
    return fig


def _update_bar_polar(fig):
    fig.update_traces(showlegend=False)
    fig.update_polars(radialaxis_showticklabels=True)
    fig.update_layout(
        polar_radialaxis_ticksuffix="",
        polar_angularaxis_rotation=90,
        polar_angularaxis_direction="clockwise",
        polar_radialaxis_dtick=10,
        polar_hole=0.10,
    )
    return fig


hierarchical_review_mx/heat_map_consistent_altair.py
import altair as alt


def make_heat_map_of_sonsistent(data, team, color):
    played_minutes = data[data.team == team]
    chart = (
        alt.Chart(played_minutes, title=f"Minutes Played by Player and Match: \n{team}")
        .mark_rect()
        .encode(
            alt.X("match:N", sort=alt.EncodingSortField(field="date", order="ascending")).title(
                "Match"
            ),
            alt.Y(
                "player:N",
                sort=alt.EncodingSortField(field="minutes", op="sum", order="descending"),
                title="Player",
            ),
            alt.Color("minutes:Q", scale=alt.Scale(scheme=color)).title("Minutes"),
            tooltip=[
                alt.Tooltip("match:N", title="Match"),
                alt.Tooltip("player:N", title="Player"),
                alt.Tooltip("minutes:Q", title="Minutes"),
            ],
        )
        .configure_view(step=13, strokeWidth=0)
        .configure_axis(domain=False, labelFontSize=10)
    )
    return chart


streamlit_app.py
import altair as alt
import pandas as pd
import plotly.express as px
import streamlit as st
import streamlit_nies as sn


larga = pd.read_csv("static/larga_player.csv")
data = pd.read_csv("static/played_minutes.csv")
# ----------------- game start --------
radar_player = "J. Musiala"

fig = sn.make_bar_plot_player(larga, radar_player)

league, team, player = st.tabs(["League", "Team", "Player"])

with league:
    st.subheader("Gráficas de desempeño")
    """
    Estas gráficas tienen un conjunto de métricas seleccionadas a partir de técnicas de inteligencia artificial.
    Cada barra representa la fuerza relativa del jugador en cada una de las métricas.
    La distancia que existe de la barra al centro indica el percentil comparado con la base de datos completa.

    La descripción completa la encontrarás en la entrada [Gráfica de desempeño de jugadores](https://www.nies.futbol/2023/07/grafica-de-desempeno-de-jugadores.html).
    """
    st.plotly_chart(fig)

with team:
    st.subheader("Gráficas de consistencia")
    """
    En la figura de abajo mostramos un mapa de calor.
    En los renglones podemos ver a los jugadores del equipo (incluyendo a los sustitutos).
    Las columnas corresponden a los partidos disputados.
    Así, el color de cada cuadro representa los minutos disputados en un partido por cada jugador.

    La descripción completa la encontrarás en la entrada [Consistencia en las alineaciones](https://www.nies.futbol/2023/08/consistencia-en-las-alineaciones-la.html).
    """
    teams = ["Cimarrones", "Cancún", "Mineros de Zacatecas"]
    colours = {"Cimarrones": "oranges", "Cancún": "blues", "Mineros de Zacatecas": "reds"}
    team = st.selectbox("Selecciona un equipo:", teams)
    color = colours[team]
    played_minutes = data[data.team == team]

    # Crear el gráfico de Altair
    hm_consistent = sn.make_heat_map_of_sonsistent(data, team, color)
    st.altair_chart(hm_consistent)

with player:
    st.subheader("Gráficas de desempeño")
    """
    Estas gráficas tienen un conjunto de métricas seleccionadas a partir de técnicas de inteligencia artificial.
    Cada barra representa la fuerza relativa del jugador en cada una de las métricas.
    La distancia que existe de la barra al centro indica el percentil comparado con la base de datos completa.

    La descripción completa la encontrarás en la entrada [Gráfica de desempeño de jugadores](https://www.nies.futbol/2023/07/grafica-de-desempeno-de-jugadores.html).
    """
    fig = sn.add_nies_logo(fig)
    st.plotly_chart(fig)


st.markdown("Made with 💖 by [nies.futbol](https://nies.futbol)")
